qa/L0_response_stats/response_stats_test.py

```diff
@@ -0,0 +1,157 @@
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from collections import defaultdict
+import sys
+
+sys.path.append("../common")
+
+import tritonclient.grpc as grpcclient
+import tritonclient.http as httpclient
+import test_util as tu
+import queue
+import numpy as np
+from functools import partial
+from tritonclient.utils import InferenceServerException
+import unittest
+
+
+class UserData:
+
+    def __init__(self):
+        self._completed_requests = queue.Queue()
+
+
+def callback(user_data, result, error):
+    if error:
+        user_data._completed_requests.put(error)
+    else:
+        user_data._completed_requests.put(result)
+
+
+class ResponseStatsTest(tu.TestResultCollector):
+
+    def setUp(self):
+        # We can only use the GRPC streaming interface because we are testing a
+        # decoupled model.
+        self._client = grpcclient.InferenceServerClient("localhost:8001")
+        self._model_name = 'square_int32'
+        self._user_data = UserData()
+        self._client.start_stream(callback=partial(callback, self._user_data))
+        self._http_client = httpclient.InferenceServerClient('localhost:8000')
+
+    def _wait_until_responses_complete(self, number_of_responses):
+        user_data = self._user_data
+        recv_count = 0
+        while recv_count < number_of_responses:
+            data_item = user_data._completed_requests.get()
+            if type(data_item) == InferenceServerException:
+                raise data_item
+
+            recv_count += 1
+
+    def _send_request(self, number_of_responses):
+        value_data = np.array([number_of_responses], dtype=np.int32)
+        inputs = []
+        inputs.append(grpcclient.InferInput('IN', value_data.shape, "INT32"))
+
+        inputs[0].set_data_from_numpy(value_data)
+        outputs = []
+        outputs.append(grpcclient.InferRequestedOutput('OUT'))
+
+        self._client.async_stream_infer(model_name=self._model_name,
+                                        inputs=inputs,
+                                        outputs=outputs)
+        self._wait_until_responses_complete(number_of_responses)
+
+    def _check_success_duration(self, duration, count):
+        self.assertEqual(duration['count'], count)
+        self.assertGreater(duration['ns'], 0)
```

Comment:
Does it make sense to add some custom delays in square backend that should be treated as compute_infer in generating responses?
These delays can be provided as backend parameters in the config.
We can then set it to that value instead of 0.

Reply:
Added custom delays.